ui/src/app/core/capabilities-activator.service.ts

```diff
@@ -0,0 +1,72 @@
+import {ActivatedRouteSnapshot, RouterStateSnapshot, Router} from '@angular/router';
+import {CanActivate} from '@angular/router';
+import {Injectable} from '@angular/core';
+
+import {AuthService} from './auth.service';
+import {CapabilitiesResponse} from '../shared/model/CapabilitiesResponse';
+import {CapabilitiesService} from './capabilities.service';
+import {environment} from '../../environments/environment';
+import {URLSearchParamsUtils} from '../shared/utils/url-search-params.utils'
+
+/** Service wrapper for routing based on current authentication state. */
+@Injectable()
+export class CapabilitiesActivator implements CanActivate {
+
+  constructor(
+    private readonly authService: AuthService,
+    private readonly capabilitiesService: CapabilitiesService,
+    private readonly router: Router) {}
+
+  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Promise<boolean> {
+    return this.capabilitiesService.getCapabilities()
+      .then(capabilities =>
+        this.handleAuthCapabilities(
+          capabilities, route.routeConfig.path, state.url))
+      .then(result => result[0] == false
+          ? Promise.resolve([false, result[1]])
+          : this.handleProjectCapabilities(result[1], route))
+      .then(result => result[0] as boolean);
+  }
+
+  private handleAuthCapabilities(capabilities: CapabilitiesResponse, path: String, url: String): [boolean, CapabilitiesResponse]|Promise<[boolean, CapabilitiesResponse]> {
+    if (capabilities.authentication && capabilities.authentication.isRequired) {
+      if (this.authService.authenticated.getValue() || path == 'sign_in') {
+        return [true, capabilities];
+      }
+
+      return this.authService.isAuthenticated().then( (authenticated) => {
+        if (!authenticated) {
+          this.router.navigate(['sign_in'], {
+            queryParams: { returnUrl: url }
+          });
+        }
+        return [authenticated, capabilities] as [boolean, CapabilitiesResponse];
```

Comment:
wdyt about erroring the promise instead in the unauthenticated case, e.g. by throwing an error here, or returning a `Promise.error('foo')`

Reply:
I'd probably prefer not too, why you think that'd be better than just returning false for the `canActivate` and navigating to the sign in page. IMO navigating to a page you are not yet authenticated for isn't really an error, it just triggers a different application flow to authenticate you before returning to the page you were on.